Add filesystem tutorial (#1775)
dlt-hub/dlt: docs/website/docs/tutorial/filesystem.md

---
title: Load data from Cloud Storage or a local filesystem
description: Load data from
keywords: [tutorial, filesystem, cloud storage, dlt, python, data pipeline, incremental loading]
---

This tutorial is for you if you need to load data files like `jsonl`, `csv`, `parquet` from
either Cloud Storage (ex. AWS S3, Google Cloud Storage, Google Drive, Azure) or a local filesystem.

## What you will learn

- How to set up a filesystem or cloud storage as a data source
- Configuration basics for filesystems and cloud storage
- Loading methods
- Incremental loading of data from filesystems or cloud storage
- How to load data of any type

## 0. Prerequisites

- Python 3.9 or higher installed
- Virtual environment set up
- `dlt` installed. Follow the instructions in the [installation guide](../reference/installation) to create a new virtual environment and install dlt.

## 1. Setting up a new project

To help you get started quickly, `dlt` provides some handy CLI commands. One of these commands will help you set up a new `dlt` project:

> ```

Reviewer: ```suggestion
```sh
```

```txt
filesystem_pipeline.py
requirements.txt
.dlt/
    config.toml
    secrets.toml
```

Here’s what each file does:

- `filesystem_pipeline.py`: This is the main script where you'll define your data pipeline. It contains several different examples of loading data from filesystem source.
- `requirements.txt`: This file lists all the Python dependencies required for your project.
- `.dlt/`: This directory contains the [configuration files](../general-usage/credentials/) for your project:
    - `secrets.toml`: This file stores your API keys, tokens, and other sensitive information.
    - `config.toml`: This file contains the configuration settings for your dlt project.

:::note
When deploying your pipeline in a production environment, managing all configurations with files
might not be convenient. In this case, you can use environment variables or other [configuration providers](../general-usage/credentials/setup#available-config-providers) available in
`dlt` to store secrets and configs instead.
:::

## 2. Creating the pipeline

Filesystem source provides users with building blocks for loading data from any type of files. You can break down the data extraction into two steps:

> 1. Accessing files in the bucket / directory.

Author: don't understand why chatgpt didn't fix it :c

2. Reading the files and yielding records.

`dlt`'s filesystem source includes several resources that you can use together or individually:

- `filesystem` resource accesses files in the directory or bucket
- several readers resources (`read_csv`, `read_parquet`, `read_jsonl`) read files and yield the records

Let's initialize a source and create a pipeline for loading parquet files from an AWS S3 to DuckDB:

```
import dlt
from dlt.sources.filesystem import filesystem, read_parquet

parquet_resource = filesystem(file_glob="**/*.parquet") | read_parquet()
pipeline = dlt.pipeline(pipeline_name="s3_to_duckdb", dataset_name="parquet_data", destination="duckdb")
info = pipeline.run(parquet_resource)
print(info)
```

You can replace the example code from `filesystem_pipeline.py` with the code above. What's happening in this snippet?

1. We initialize the filesystem resource and pass the `file_glob` parameter. Based in this parameter, `dlt` will filter
all files in the bucket.

> 2. We pipe the transformer `read_parquet` to read the files yielded by filesystem resource and iterate over records

Reviewer 2: Would suggest to rephrase to something like: We use a pipe to pass files yielded into the `read_parquet` transformer, which will read the file and iterate over its records.

from the file.

  ```note
  A **transformer** in `dlt` is a special type of resource that processes each record from another resource. This lets you
  chain multiple resources together. To learn more, check out the
  ["@dlt.transformer section"](../general-usage/resource#process-resources-with-dlttransformer).
  ```

> 3. We create the `dlt` pipeline with the name `s3_to_duckdb` and DuckDB destination and run this pipeline.

Reviewer 2: ```suggestion
3. We create the `dlt` pipeline with the name `s3_to_duckdb` and a DuckDB destination and run the pipeline.
```

> ## 3. Configuring filesystem source

Reviewer 2: ```suggestion
## 3. Configuring the filesystem source
```

Next, we need to configure the connection. Specifically, we’ll set the bucket URL and AWS credentials.
This example focuses on AWS S3. For other Cloud Storage services, see the [Filesystem configuration section](../dlt-ecosystem/verified-sources/filesystem/basic#configuration).

> Let's specify the bucket url and AWS credentials:

Reviewer 2: are you sure we should include all three options of specifying the credentials? In the earlier example we only have the toml example

[Configuration and secrets section](../../../general-usage/credentials/complex_types#aws-credentials).
:::

## 4. Running the pipeline

Let's verify that the pipeline is working as expected. Run the following command to execute the pipeline:

```sh
python filesystem_pipeline.py
```

You should see the output of the pipeline execution in the terminal. The output will also display the location of the DuckDB database file where the data is stored:

```sh
Pipeline s3_to_duckdb load step completed in 1.08 seconds
1 load package(s) were loaded to destination duckdb and into dataset parquet_data
The duckdb destination used duckdb:////home/user-name/quick_start/s3_to_duckdb.duckdb location to store data
Load package 1692364844.3523829 is LOADED and contains no failed jobs
```

## 5. Exploring the data

Now that the pipeline has run successfully, let's explore the data loaded into DuckDB. `dlt` comes with a built-in browser application that allows you to interact with the data. To enable it, run the following command:

```sh
pip install streamlit
```

Next, run the following command to start the data browser:

```sh
dlt pipeline s3_to_duckdb show
```

> The command opens a new browser window with the data browser application. `s3_to_duckdb` is the name of the pipeline defined in the `filesystem_pipeline.py` file.

Reviewer 2: should there be a screenshot here?

You can explore the loaded data, run queries and see some pipeline execution details.

## 6. Appending, replacing, and merging loaded data

Try running the pipeline again with `python filesystem_pipeline.py`. You will notice that

> all the tables have data duplicated. This happens because by default, `dlt` appends the data to the destination table. It is very useful, for example, when you have daily data updates and you want to ingest them. In `dlt` you can control how the data is loaded into the destination table by setting the `write_disposition` parameter in the resource configuration. The possible values are:

Reviewer 2: ```suggestion
all the tables have duplicate data. This happens because by default, `dlt` appends the data to the destination table. It is very useful, for example, when you have daily data updates and you want to ingest them. In `dlt` you can control how the data is loaded into the destination table by setting the `write_disposition` parameter in the resource configuration. The possible values are:
```

Reviewer 2: I left some comments about this paragraph on Anton's PR as well -  we should coordinate that.

- `append`: Appends the data to the destination table. This is the default.
- `replace`: Replaces the data in the destination table with the new data.
- `merge`: Merges the new data with the existing data in the destination table based on the primary key.

> To specify the `write_disposition` you can set it in `pipeline.run` command:

Reviewer 2: ```suggestion
To modify the `write_disposition` you can set it in the `pipeline.run` command:
```

```py
import dlt
from dlt.sources.filesystem import filesystem, read_parquet

parquet_resource = filesystem(file_glob="**/*.parquet") | read_parquet()
pipeline = dlt.pipeline(pipeline_name="s3_to_duckdb", dataset_name="parquet_data", destination="duckdb")
info = pipeline.run(parquet_resource, write_disposition="replace")
print(info)
```

> You can learn more about write_disposition in the [Write dispositions section](../general-usage/incremental-loading#the-3-write-dispositions).

Reviewer 2: Ok so in Anton's tutorial he added examples for each of the three options - do we not want to do that here?

## 7. Loading data incrementally

When loading data from files, you often only want to load files that have been modified. `dlt` makes this easy with [incremental loading](../general-usage/incremental-loading). To load only modified files you can use `apply_hint`
function:

```py
import dlt
from dlt.sources.filesystem import filesystem, read_parquet

files_resource = filesystem(file_glob="**/*.parquet")
files_resource.apply_hints(incremental=dlt.sources.incremental("modification_date"))
parquet_resource = files_resource | read_parquet()
pipeline = dlt.pipeline(pipeline_name="s3_to_duckdb", dataset_name="parquet_data", destination="duckdb")
info = pipeline.run(parquet_resource, write_disposition="replace")
print(info)
```

As you can see the incremental hint is applied to the `filesystem` resource before `read_parquet` transformer, because

> we want to apply the incremental loading to filter the loaded files.

Reviewer 2: this is also less verbose than Anton's tutorial - should we include the full instructions?

Let's create and apply a transformer, which reads `json` files instead of `parquet` (the implementation for `json`
is a little bit different than for `jsonl`).

```
import dlt
from dlt.common.storages.fsspec_filesystem import FileItemDict
from dlt.common.typing import TDataItems
from dlt.sources.filesystem import filesystem

# Define a standalone transformer to read data from an json file.
@dlt.transformer(standalone=True)
def read_json(items: Iterator[FileItemDict]) -> Iterator[TDataItems]:
    for file_obj in items:
        with file_obj.open() as f:
            yield json.load(f)

files_resource = filesystem(file_glob="**/*.json")
files_resource.apply_hints(incremental=dlt.sources.incremental("modification_date"))
json_resource = files_resource | read_json()
pipeline = dlt.pipeline(pipeline_name="s3_to_duckdb", dataset_name="json_data", destination="duckdb")
info = pipeline.run(json_resource, write_disposition="replace")
print(info)
```

Check out [other examples](../dlt-ecosystem/verified-sources/filesystem/advanced#create-your-own-transformer) showing how to read data from `excel` and `xml` files.

## What's next?

> Congratulations on completing the tutorial! You've learned how to set up a Filesystem source in `dlt` and run a data pipeline to load the data into DuckDB.

Reviewer 2: ```suggestion
Congratulations on completing the tutorial! You've learned how to set up a filesystem source in `dlt` and run a data pipeline to load the data into DuckDB.
```

Interested in learning more about `dlt`? Here are some suggestions:

> - Learn more about the Filesystem source configuration in [Filesystem source](../dlt-ecosystem/verified-sources/filesystem)

Reviewer 2: ```suggestion
- Learn more about the filesystem source configuration in [the filesystem source reference](../dlt-ecosystem/verified-sources/filesystem)
```

## 3. Configuring filesystem source

:::note
In this tutorial we will work with publicly accessed dataset [Hospital Patient Records](https://mavenanalytics.io/data-playground?order=date_added%2Cdesc&search=Hospital%20Patient%20Records)
synthetic electronic health care records. You can use the exact credentials from this tutorial to load this dataset from GCP.
<details>
<summary>Citation</summary>
Jason Walonoski, Mark Kramer, Joseph Nichols, Andre Quina, Chris Moesel, Dylan Hall, Carlton Duffett, Kudakwashe Dube, Thomas Gallagher, Scott McLachlan, Synthea: An approach, method, and software mechanism for generating synthetic patients and the synthetic electronic health care record, Journal of the American Medical Informatics Association, Volume 25, Issue 3, March 2018, Pages 230–238, https://doi.org/10.1093/jamia/ocx079
</details>
:::

Next, we need to configure the connection. Specifically, we’ll set the bucket URL and credentials. This example focuses on Google Cloud Storage. For other Cloud Storage services, see the [Filesystem configuration section](../dlt-ecosystem/verified-sources/filesystem/basic#configuration).

Let's specify the bucket URL and credentials:

> <Tabs

Reviewer: I think for the tutorial we can just pick one type of configuration (the most common/convenient for the tutorial) and just link the docs for alternative ways.

Author: I thought this two at first, but one of the users complained to me that we have examples with toml files everywhere and it's difficult to get other examples. I think it doesn't hurt to have options, wdyt?

- the `filesystem` resource accesses files in the directory or bucket
- several readers resources (`read_csv`, `read_parquet`, `read_jsonl`) read files and yield the records. These resources have a
special type, they called [transformers](../general-usage/resource#process-resources-with-dlttransformer). Transformers expect items from another resource.
In this particular case transformers expect `FileItem` object and transform it into multiple records.

Let's initialize a source and create a pipeline for loading `csv` files from Google Cloud Storage to DuckDB. You can replace code from `filesystem_pipeline.py` with the following:

```py
import dlt
from dlt.sources.filesystem import filesystem, read_csv

files = filesystem(bucket_url="gs://filesystem-tutorial", file_glob="encounters*.csv")
reader = (files | read_csv()).with_name("encounters")
pipeline = dlt.pipeline(pipeline_name="hospital_data_pipeline", dataset_name="hospital_data", destination="duckdb")
info = pipeline.run(reader)
print(info)
```

What's happening in this snippet?

1. We initialize the filesystem resource and pass the `file_glob` parameter. Based on this parameter, `dlt` will filter all files in the bucket.

> 2. We pipe the transformer `read_csv` to read the files yielded by the filesystem resource and iterate over records from the file.

Reviewer: ```suggestion
2. We pipe the files names yielded by the filesystem resource to the transformer resource `read_csv` to read each file and iterate over records from the file. We name this transformer resource `"encounters"` using the `with_name()`. dlt will use the resource name `"encounters"` as a table name when loading the data.
```
Add more details

:::note

> A **transformer** in `dlt` is a special type of resource that processes each record from another resource. This lets you chain multiple resources together. To learn more, check out the [transformer section](../general-usage/resource#process-resources-with-dlttransformer).

Reviewer: ```suggestion
A [transformer](../general-usage/resource#process-resources-with-dlttransformer) in `dlt` is a special type of resource that processes each record from another resource. This lets you chain multiple resources together.
```

:::

> 3. We create the `dlt` pipeline with the name `hospital_data_pipeline` and DuckDB destination and run this pipeline.

Reviewer: ```suggestion
3. We create the `dlt` pipeline configuring with the name `hospital_data_pipeline` and DuckDB destination.
4. We call `pipeline.run()`. This is where the underlying generators are iterated:
 - dlt retrieves remote data,
 - normalizes data,
 - creates or updates the table in the destination,
 - loads the extracted data into the destination.
 5. `print(info)` outputs pipeline running stats we get from `pipeline.run()`
```

## 3. Configuring the filesystem source

:::note
In this tutorial we will work with publicly accessed dataset [Hospital Patient Records](https://mavenanalytics.io/data-playground?order=date_added%2Cdesc&search=Hospital%20Patient%20Records)
synthetic electronic health care records. You can use the exact credentials from this tutorial to load this dataset from GCP.
<details>
<summary>Citation</summary>
Jason Walonoski, Mark Kramer, Joseph Nichols, Andre Quina, Chris Moesel, Dylan Hall, Carlton Duffett, Kudakwashe Dube, Thomas Gallagher, Scott McLachlan, Synthea: An approach, method, and software mechanism for generating synthetic patients and the synthetic electronic health care record, Journal of the American Medical Informatics Association, Volume 25, Issue 3, March 2018, Pages 230–238, https://doi.org/10.1093/jamia/ocx079
</details>
:::

> Next, we need to configure the connection. Specifically, we’ll set the bucket URL and credentials. This example focuses on Google Cloud Storage. For other Cloud Storage services, see the [Filesystem configuration section](../dlt-ecosystem/verified-sources/filesystem/basic#configuration).

Reviewer: ```suggestion
Next, we need to configure the connection. Specifically, we’ll set the bucket URL and credentials. This example uses Google Cloud Storage. For other cloud storage services, see the [Filesystem configuration section](../dlt-ecosystem/verified-sources/filesystem/basic#configuration).
```

> Let's specify the bucket URL and credentials:

Reviewer: ```suggestion
Let's specify the bucket URL and credentials. We can do this using the following methods:
```

![Streamlit Explore data](/img/filesystem-tutorial/streamlit-data.png)

You can explore the loaded data, run queries, and see some pipeline execution details.

## 6. Appending, replacing, and merging loaded data

If you try running the pipeline again with `python filesystem_pipeline.py`, you will notice that all the tables have duplicated data. This happens because by default, `dlt` appends the data to the destination table. It is very useful, for example, when you have daily data updates and you want to ingest them. With `dlt`, you can control how the data is loaded into the destination table by setting the `write_disposition` parameter in the resource configuration. The possible values are:
- `append`: Appends the data to the destination table. This is the default.
- `replace`: Replaces the data in the destination table with the new data.
- `merge`: Merges the new data with the existing data in the destination table based on the primary key.

To specify the `write_disposition`, you can set it in the `pipeline.run` command. Let's change the write disposition to `merge`. In this case, `dlt` will deduplicate the data before loading them into the destination.

> To enable data deduplication, we also should specify a `primary_key` or `merge_key`, which will be used by `dlt` to define if two records are different. Both keys could consist of several columns. `dlt` will try to use `merge_key` and fallback to `primary_key` if it's not specified. To specify any hints about the data, including column types, primary keys, you can use the [`apply_hints` function](../general-usage/resource#set-table-name-and-adjust-schema).

Reviewer: ```suggestion
To enable data deduplication, we also should specify a `primary_key` or `merge_key`, which will be used by `dlt` to define if two records are different. Both keys could consist of several columns. `dlt` will try to use `merge_key` and fallback to `primary_key` if it's not specified. To specify any hints about the data, including column types, primary keys, you can use the [`apply_hints`](../general-usage/resource#set-table-name-and-adjust-schema) method.
```
Formatting

```py
import dlt
from dlt.sources.filesystem import filesystem, read_csv

files = filesystem(file_glob="encounters*.csv")
reader = (files | read_csv()).with_name("encounters")
reader.apply_hints(primary_key="id")

> pipeline = dlt.pipeline(pipeline_name="hospital_data_pipeline", dataset_name="hospital_data", destination="duckdb")

Reviewer: ```suggestion
pipeline = dlt.pipeline(pipeline_name="hospital_data_pipeline", dataset_name="hospital_data", destination="duckdb")

```
Add some vertical space

info = pipeline.run(reader, write_disposition="merge")
print(info)
```
:::tip
You may need to drop the previously loaded data if you loaded data several times with `append` write disposition to make sure the primary key column has unique values.
:::

> You can learn more about `write_disposition` in the [Write dispositions section](../general-usage/incremental-loading#the-3-write-dispositions).

Reviewer: ```suggestion
You can learn more about `write_disposition` in the [write dispositions section](../general-usage/incremental-loading#the-3-write-dispositions) of the incremental loading page.
```

## 7. Loading data incrementally

> When loading data from files, you often only want to load files that have been modified. `dlt` makes this easy with [incremental loading](../general-usage/incremental-loading). To load only modified files, you can use the `apply_hint` function:

Reviewer: ```suggestion
When loading data from files, you often only want to load files that have been modified. `dlt` makes this easy with [incremental loading](../general-usage/incremental-loading). To load only modified files, you can use the `apply_hint` method:
```

```py
import dlt
from dlt.sources.filesystem import filesystem, read_csv

files = filesystem(file_glob="encounters*.csv")
files.apply_hints(incremental=dlt.sources.incremental("modification_date"))
reader = (files | read_csv()).with_name("encounters")
reader.apply_hints(primary_key="id")

> pipeline = dlt.pipeline(pipeline_name="hospital_data_pipeline", dataset_name="hospital_data", destination="duckdb")

Reviewer: ```suggestion
pipeline = dlt.pipeline(pipeline_name="hospital_data_pipeline", dataset_name="hospital_data", destination="duckdb")

```
Add vertical space

info = pipeline.run(reader, write_disposition="merge")
print(info)
```

Notice that we used `apply_hints` on the `files` resource, not on `reader`. Why did we do that? As mentioned before, the `filesystem` resource lists all files in the storage based on the `file_glob` parameter. So at this point, we can also specify additional conditions to filter out files. In this case, we only want to load files that have been modified since the last load. `dlt` will automatically keep the state of incremental load and manage the correct filtering.

> But what if we not only want to process modified files, but we also want to load only new records? In the `encounters` table, we can see the column named `STOP` indicating the timestamp of the end of the encounter. Let's modify our code to load only those records whose `STOP` timestamp was updated since our last load.

Reviewer: This is an awesome example! 🔥

```py
import dlt
from dlt.sources.filesystem import filesystem, read_csv

files = filesystem(file_glob="encounters*.csv")
files.apply_hints(incremental=dlt.sources.incremental("modification_date"))
reader = (files | read_csv()).with_name("encounters")
reader.apply_hints(primary_key="id", incremental=dlt.sources.incremental("STOP"))

> pipeline = dlt.pipeline(pipeline_name="hospital_data_pipeline", dataset_name="hospital_data", destination="duckdb")

Reviewer: ```suggestion
pipeline = dlt.pipeline(pipeline_name="hospital_data_pipeline", dataset_name="hospital_data", destination="duckdb")

```
Add vertical spacing

```py
from typing import Any, Iterator

import dlt
from dlt.sources import TDataItems
from dlt.sources.filesystem import FileItemDict
from dlt.sources.filesystem import filesystem

@dlt.transformer()
def read_csv_custom(items: Iterator[FileItemDict], chunksize: int = 10000, **pandas_kwargs: Any) -> Iterator[TDataItems]:
    import pandas as pd

    # apply defaults to pandas kwargs
    kwargs = {**{"header": "infer", "chunksize": chunksize}, **pandas_kwargs}

    for file_obj in items:
        with file_obj.open() as file:
            for df in pd.read_csv(file, **kwargs):
                df["file_name"] = file_obj["file_name"]
                yield df.to_dict(orient="records")

files = filesystem(file_glob="encounters*.csv")
files.apply_hints(incremental=dlt.sources.incremental("modification_date"))
reader = (files | read_csv_custom()).with_name("encounters")
reader.apply_hints(primary_key="id", incremental=dlt.sources.incremental("STOP"))

> pipeline = dlt.pipeline(pipeline_name="hospital_data_pipeline", dataset_name="hospital_data", destination="duckdb")

Reviewer: ```suggestion
pipeline = dlt.pipeline(pipeline_name="hospital_data_pipeline", dataset_name="hospital_data", destination="duckdb")

```
Add vertical spacing

Let's create and apply a transformer that reads `json` files instead of `csv` (the implementation for `json` is a little bit different from `jsonl`).

```py
from typing import Iterator

import dlt
from dlt.common import json
from dlt.common.storages.fsspec_filesystem import FileItemDict
from dlt.common.typing import TDataItems
from dlt.sources.filesystem import filesystem

# Define a standalone transformer to read data from a json file.
@dlt.transformer(standalone=True)
def read_json(items: Iterator[FileItemDict]) -> Iterator[TDataItems]:
    for file_obj in items:
        with file_obj.open() as f:
            yield json.load(f)

files_resource = filesystem(file_glob="**/*.json")
files_resource.apply_hints(incremental=dlt.sources.incremental("modification_date"))
json_resource = files_resource | read_json()

> pipeline = dlt.pipeline(pipeline_name="s3_to_duckdb", dataset_name="json_data", destination="duckdb")

Reviewer: ```suggestion
pipeline = dlt.pipeline(pipeline_name="s3_to_duckdb", dataset_name="json_data", destination="duckdb")

```
Add vertical spacing
